pkg/kv/store_message.go

```diff
@@ -51,62 +51,133 @@ func (s *StoreMessage) DeleteMsg(ctx context.Context, path string) error {
-func (s *StoreMessage) Scan(ctx context.Context, msgType protoreflect.MessageType, prefix string) (*MessageIterator, error) {
-	return NewMessageIterator(ctx, s.Store, msgType, prefix)
-}
-
-type MessageIterator struct {
-	itr     EntriesIterator
-	msgType protoreflect.MessageType
-	err     error
+func (s *StoreMessage) Scan(ctx context.Context, prefix string) (*PrimaryIterator, error) {
+	return NewPrimaryIterator(ctx, s.Store, prefix)
 }
 
 type MessageEntry struct {
 	Key   string
 	Value protoreflect.ProtoMessage
 }
 
-func NewMessageIterator(ctx context.Context, store Store, msgType protoreflect.MessageType, prefix string) (*MessageIterator, error) {
+func NewMessageEntry(msg protoreflect.MessageType) *MessageEntry {
+	return &MessageEntry{
+		Key:   "",
+		Value: msg.New().Interface(),
+	}
+}
+
+type MessageIterator interface {
+	Next() bool
+	Entry(key *string, value *protoreflect.ProtoMessage) error
+	Err() error
+	Close()
+}
+
+// PrimaryIterator MessageIterator implementation for primary key
+// The iterator iterates over the given prefix and returns the proto message and key
+type PrimaryIterator struct {
+	itr EntriesIterator
+	err error
+}
+
+func NewPrimaryIterator(ctx context.Context, store Store, prefix string) (*PrimaryIterator, error) {
 	itr, err := ScanPrefix(ctx, store, []byte(prefix))
 	if err != nil {
 		return nil, fmt.Errorf("create prefix iterator: %w", err)
 	}
-	return &MessageIterator{itr: itr, msgType: msgType}, nil
+	return &PrimaryIterator{itr: itr}, nil
 }
 
-func (m *MessageIterator) Next() bool {
+func (m *PrimaryIterator) Next() bool {
 	if m.err != nil {
 		return false
 	}
 	return m.itr.Next()
 }
 
-func (m *MessageIterator) Entry() *MessageEntry {
+func (m *PrimaryIterator) Entry(key *string, value *protoreflect.ProtoMessage) error {
 	entry := m.itr.Entry()
 	if entry == nil {
-		return nil
+		return ErrNotFound
 	}
-	msg := m.msgType.New().Interface()
-	err := proto.Unmarshal(entry.Value, msg)
-	if err != nil {
-		m.err = fmt.Errorf("unmarshal proto data for key %s: %w", string(entry.Key), err)
-		return nil
+	if value != nil {
+		err := proto.Unmarshal(entry.Value, *value)
+		if err != nil {
+			return fmt.Errorf("unmarshal proto data for key %s: %w", string(entry.Key), err)
+		}
 	}
-	return &MessageEntry{
-		Key:   string(entry.Key),
-		Value: msg,
+	if key != nil {
+		*key = string(entry.Key)
 	}
+	return nil
 }
 
-func (m *MessageIterator) Err() error {
+func (m *PrimaryIterator) Err() error {
 	if m.err != nil {
 		return m.err
 	}
 	return m.itr.Err()
 }
 
-func (m *MessageIterator) Close() {
+func (m *PrimaryIterator) Close() {
 	m.itr.Close()
 }
+
+// SecondaryIterator MessageIterator implementation for secondary key
+// The iterator iterates over the given prefix, extracts the primary key value from secondary key and then returns
+// the proto message and primary key
+type SecondaryIterator struct {
```

Comment:
If we'll use proto for secondary indexes, we won't have to create special iterators for them.

Reply:
I understand your comment about defining a protobuf for secondary indexes but I think we will still have to create a special iterator for secondary indices since the Entry functionality is different.
Maybe we can discuss it further?